lib/master.js

```diff
@@ -218,24 +230,26 @@ class Master extends BaseService {
                     worker_pid: worker.process.pid,
                     exit_code: worker.process.exitCode
                 };
-                if (this.workerStatusMap[worker.process.pid] &&
-                    this.workerStatusMap[worker.process.pid].status) {
-                    info.status = this.workerStatusMap[worker.process.pid].status;
+                if (this._workerStatusMap[worker.process.pid] &&
+                    this._workerStatusMap[worker.process.pid].status) {
+                    info.status = this._workerStatusMap[worker.process.pid].status;
                 }
                 this._logger.log('error/service-runner/master', info);
-                delete this.workerStatusMap[worker.process.pid];
+                delete this._workerStatusMap[worker.process.pid];
+                worker.removeListener('exit', workerExitHandler);
+                worker.removeListener('message', workerMessageHandler);
+
                 P.delay(Math.random() * 2000).then(() => {
                     resolve(this._startWorker(worker.worker_id));
                 });
             };
 
-            worker.on('exit', startupWorkerExit);
-            worker.on('message', (msg) => {
+            workerMessageHandler = (msg) => {
                 switch (msg.type) {
                     case 'startup_finished':
-                        worker.removeListener('exit', startupWorkerExit);
+                        worker.removeListener('exit', startupWorkerExitHandler);
                         worker.on('exit', () => {
-                            workerExit(worker);
+                            workerExitHandler(worker);
```

Comment:
No action needed. TODO: mental note for myself - we can use the worker in the exit handler from the closure and pass in the exit code to log it as well.